File: app/lib/models/booking.dart
class Booking {
  final String id;
  final String gameId;
  final String playerId;
  final String status;
  final DateTime createdAt;

  const Booking({
    required this.id,
    required this.gameId,
    required this.playerId,
    this.status = 'confirmed',
    required this.createdAt,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['id'] as String,
      gameId: json['game_id'] as String,
      playerId: json['player_id'] as String,
      status: json['status'] as String? ?? 'confirmed',
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'game_id': gameId,
        'player_id': playerId,
        'status': status,
        'created_at': createdAt.toIso8601String(),
      };
}

File: api/models/booking.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import ForeignKey, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class Booking(Base):
    __tablename__ = "bookings"

    id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    game_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("games.id"), nullable=False
    )
    player_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    status: Mapped[str] = mapped_column(
        String(20), default="confirmed"
    )  # invited, confirmed, waitlisted, declined, cancelled
    # Only `confirmed` occupies a slot and counts toward Game.current_players.
    # `invited` is awaiting a response; `declined`/`cancelled` hold nothing.
    created_at: Mapped[str] = mapped_column(
        server_default=func.now(),
        default=lambda: datetime.now(timezone.utc).isoformat(),
    )

    __table_args__ = (
        UniqueConstraint("game_id", "player_id", name="uq_booking_game_player"),
    )

